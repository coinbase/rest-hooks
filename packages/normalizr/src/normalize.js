import EntitySchema from './schemas/Entity';
import UnionSchema from './schemas/Union';
import ValuesSchema from './schemas/Values';
import ArraySchema, * as ArrayUtils from './schemas/Array';
import ObjectSchema, * as ObjectUtils from './schemas/Object';

const visit = (value, parent, key, schema, addEntity, visitedEntities) => {
  if (typeof value !== 'object' || !value || !schema) {
    return value;
  }

  if (
    typeof schema === 'object' &&
    (!schema.normalize || typeof schema.normalize !== 'function')
  ) {
    const method = Array.isArray(schema)
      ? ArrayUtils.normalize
      : ObjectUtils.normalize;
    return method(
      schema,
      value,
      parent,
      key,
      visit,
      addEntity,
      visitedEntities,
    );
  }

  return schema.normalize(
    value,
    parent,
    key,
    visit,
    addEntity,
    visitedEntities,
  );
};

const addEntities = (entities, indexes) => (
  schema,
  processedEntity,
  value,
  parent,
  key,
) => {
  const schemaKey = schema.key;
  const id = schema.getId(value, parent, key);
  if (!(schemaKey in entities)) {
    entities[schemaKey] = {};
  }

  const existingEntity = entities[schemaKey][id];
  if (existingEntity) {
    entities[schemaKey][id] = schema.merge(existingEntity, processedEntity);
  } else {
    entities[schemaKey][id] = processedEntity;
  }
  // update index
  if (Array.isArray(schema.indexes)) {
    const entity = entities[schemaKey][id];
    if (!(schemaKey in indexes)) {
      indexes[schemaKey] = {};
    }
    for (const index of schema.indexes) {
      if (!(index in indexes[schemaKey])) {
        indexes[schemaKey][index] = {};
      }
      const indexMap = indexes[schemaKey][index];
      if (existingEntity) {
        delete indexMap[existingEntity[index]];
      }
      if (index in entity) {
        indexMap[entity[index]] = id;
      } /* istanbul ignore next */ else if (
        // eslint-disable-next-line no-undef
        process.env.NODE_ENV !== 'production'
      ) {
        console.warn(`Index not found in entity. Indexes must be top-level members of your entity.
Index: ${index}
Entity: ${JSON.stringify(entity, undefined, 2)}`);
      }
    }
  }
};

export const schema = {
  Array: ArraySchema,
  Entity: EntitySchema,
  Object: ObjectSchema,
  Union: UnionSchema,
  Values: ValuesSchema,
};

function expectedSchemaType(schema) {
  return ['object', 'function'].includes(typeof schema)
    ? 'object'
    : typeof schema;
}

export const normalize = (input, schema) => {
  const schemaType = expectedSchemaType(schema);
  if (input === null || typeof input !== schemaType) {
    throw new Error(
      `Unexpected input given to normalize. Expected type to be "${schemaType}", found "${
        input === null ? 'null' : typeof input
      }".`,
    );
  }

  const entities = {};
  const indexes = {};
  const addEntity = addEntities(entities, indexes);
  const visitedEntities = {};

  const result = visit(
    input,
    input,
    undefined,
    schema,
    addEntity,
    visitedEntities,
  );
  return { entities, indexes, result };
};
