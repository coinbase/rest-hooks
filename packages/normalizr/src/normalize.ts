import { normalize as arrayNormalize } from './schemas/Array';
import { normalize as objectNormalize } from './schemas/Object';
import type {
  NormalizeNullable,
  NormalizedSchema,
  Schema,
  NormalizedIndex,
} from './types';
import { DELETED } from './special';

const visit = (
  value: any,
  parent: any,
  key: any,
  schema: any,
  addEntity: any,
  visitedEntities: any,
) => {
  if (!value || !schema || !['function', 'object'].includes(typeof schema)) {
    return value;
  }

  if (!schema.normalize || typeof schema.normalize !== 'function') {
    // serializable
    if (typeof schema === 'function') {
      return new schema(value);
    }
    const method = Array.isArray(schema) ? arrayNormalize : objectNormalize;
    return method(
      schema,
      value,
      parent,
      key,
      visit,
      addEntity,
      visitedEntities,
    );
  }

  return schema.normalize(
    value,
    parent,
    key,
    visit,
    addEntity,
    visitedEntities,
  );
};

const addEntities =
  (
    entities: Record<string, any>,
    indexes: any,
    existingEntities: Record<string, any>,
    existingIndexes: any,
  ) =>
  (schema: any, processedEntity: any, value: any, parent: any, key: string) => {
    const schemaKey = schema.key;
    const id = schema.pk(value, parent, key);
    if (!(schemaKey in entities)) {
      entities[schemaKey] = {};
    }

    const existingEntity = entities[schemaKey][id];
    if (existingEntity) {
      entities[schemaKey][id] = schema.merge(existingEntity, processedEntity);
    } else {
      entities[schemaKey][id] = processedEntity;
    }
    // update index
    if (Array.isArray(schema.indexes)) {
      const entity = entities[schemaKey][id];
      if (!(schemaKey in indexes)) {
        indexes[schemaKey] = {};
      }
      for (const index of schema.indexes) {
        if (!(index in indexes[schemaKey])) {
          indexes[schemaKey][index] = {};
        }
        const indexMap = indexes[schemaKey][index];
        if (existingEntity) {
          delete indexMap[existingEntity[index]];
        }
        // entity already in cache but the index changed
        if (
          existingEntities[schemaKey] &&
          existingEntities[schemaKey][id] &&
          existingEntities[schemaKey][id][index] !== entity[index]
        ) {
          indexMap[existingEntities[schemaKey][id][index]] = DELETED;
        }
        if (index in entity) {
          indexMap[entity[index]] = id;
        } /* istanbul ignore next */ else if (
          // eslint-disable-next-line no-undef
          process.env.NODE_ENV !== 'production'
        ) {
          console.warn(`Index not found in entity. Indexes must be top-level members of your entity.
Index: ${index}
Entity: ${JSON.stringify(entity, undefined, 2)}`);
        }
      }
    }
  };

function expectedSchemaType(schema: Schema) {
  return ['object', 'function'].includes(typeof schema)
    ? 'object'
    : typeof schema;
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const normalize = <
  S extends Schema = Schema,
  E extends Record<string, Record<string, any> | undefined> = Record<
    string,
    Record<string, any>
  >,
  R = NormalizeNullable<S>,
>(
  input: any,
  schema?: S,
  existingEntities: Readonly<E> = {} as any,
  existingIndexes: Readonly<NormalizedIndex> = {},
): NormalizedSchema<E, R> => {
  // no schema means we don't process at all
  if (schema === undefined)
    return {
      entities: existingEntities,
      indexes: existingIndexes,
      result: input,
    };

  const schemaType = expectedSchemaType(schema);
  if (input === null || typeof input !== schemaType) {
    /* istanbul ignore else */
    if (process.env.NODE_ENV !== 'production') {
      const parseWorks = (input: string) => {
        try {
          return typeof JSON.parse(input) !== 'string';
        } catch (e) {
          return false;
        }
      };
      if (typeof input === 'string' && parseWorks(input)) {
        throw new Error(`Normalizing a string, but this does match schema.

Parsing this input string as JSON worked. This likely indicates fetch function did not parse
the JSON. By default, this only happens if "content-type" header includes "json".
See https://resthooks.io/docs/guides/custom-networking for more information

  Schema: ${JSON.stringify(schema, undefined, 2)}
  Input: "${input}"`);
      } else {
        throw new Error(
          `Unexpected input given to normalize. Expected type to be "${schemaType}", found "${
            input === null ? 'null' : typeof input
          }".

          Schema: ${JSON.stringify(schema, undefined, 2)}
          Input: "${input}"`,
        );
      }
    } else {
      throw new Error(
        `Unexpected input given to normalize. Expected type to be "${schemaType}", found "${
          input === null ? 'null' : typeof input
        }".`,
      );
    }
  }

  const entities: E = {} as any;
  const indexes: NormalizedIndex = {};
  const addEntity = addEntities(
    entities,
    indexes,
    existingEntities,
    existingIndexes,
  );
  const visitedEntities = {};

  const result = visit(
    input,
    input,
    undefined,
    schema,
    addEntity,
    visitedEntities,
  );
  return { entities, indexes, result };
};
