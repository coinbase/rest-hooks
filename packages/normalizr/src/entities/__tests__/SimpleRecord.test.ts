import { SimpleRecord, denormalize, Entity } from '../..';
import { normalize } from '../../normalize';

class Article extends SimpleRecord {
  readonly id: string = '';
  readonly title: string = '';
  readonly author: string = '';
  readonly content: string = '';
}
class ArticleEntity extends Entity {
  readonly id: string = '';
  readonly title: string = '';
  readonly author: string = '';
  readonly content: string = '';
  pk() {
    return this.id;
  }
}

class Pagination extends SimpleRecord {
  readonly data = ArticleEntity.fromJS();
  readonly nextPage: string = '';
  readonly lastPage: string = '';

  get title() {
    return this.data.title;
  }

  static schema = {
    data: ArticleEntity,
  };
}

describe('SimpleRecord', () => {
  it('should init', () => {
    const resource = Article.fromJS({
      id: '5',
      title: 'happy',
      author: '5',
    });
    expect(resource.title).toBe('happy');
    expect(resource.author).toBe('5');
  });

  it('should not allow init with wrong types', () => {
    Article.fromJS({
      // @ts-expect-error
      id: 5,
      title: 'happy',
      // @ts-expect-error
      author: 5,
    });
  });

  describe('normalize', () => {
    it('should normalize into plain object', () => {
      const schema = Article;
      const normalized = normalize(
        {
          id: '5',
          title: 'happy',
          author: '5',
        },
        schema,
      );
      expect(normalized.result).toEqual({
        id: '5',
        title: 'happy',
        author: '5',
      });
    });

    it('should work on nested', () => {
      const schema = Pagination;
      const normalized = normalize(
        {
          data: {
            id: '5',
            title: 'happy',
            author: '5',
          },
          lastPage: 'abc',
          nextPage: 'slksfd',
        },
        schema,
      );
      expect(normalized).toMatchSnapshot();
    });
  });

  describe('denormalize', () => {
    it('should denormalize with defaults', () => {
      const schema = Article;
      const denormalized = denormalize(
        {
          id: '5',
          title: 'happy',
          author: '5',
        },
        schema,
        {},
      );
      const article = denormalized[0];
      expect(article).toBeDefined();
      expect(article).toBeInstanceOf(Article);
      expect(article).toEqual({
        id: '5',
        title: 'happy',
        author: '5',
        content: '',
      });
      // typing of members that it has
      article?.content;
      article?.author;
      // @ts-expect-error - fails when it doesn't
      article?.bob;
    });

    it('should work with nested and default correctly', () => {
      const schema = Pagination;
      const denormalized = denormalize(
        {
          data: '5',
          nextPage: 'blob',
        },
        schema,
        {
          [ArticleEntity.key]: {
            ['5']: {
              id: '5',
              title: 'happy',
              author: '5',
            },
          },
        },
      );
      expect(denormalized[1]).toBe(true);
      const pagination = denormalized[0];
      expect(pagination).toBeDefined();
      expect(pagination).toBeInstanceOf(Pagination);
      expect(pagination).toEqual({
        data: {
          id: '5',
          title: 'happy',
          author: '5',
          content: '',
        },
        nextPage: 'blob',
        lastPage: '',
      });
      // check getters
      expect(pagination.title).toBe('happy');
      // typing of members that it has
      pagination.data.content;
      pagination.data.content;
      // @ts-expect-error - fails when it doesn't
      pagination?.bob;
    });
  });
});
