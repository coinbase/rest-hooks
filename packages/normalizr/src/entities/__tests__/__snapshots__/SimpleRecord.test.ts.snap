// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleRecord normalize should work on nested 1`] = `
Object {
  "entities": Object {
    "ArticleEntity": Object {
      "5": ArticleEntity {
        "author": "5",
        "content": "",
        "id": "5",
        "title": "happy",
      },
    },
  },
  "indexes": Object {},
  "result": Object {
    "data": "5",
    "lastPage": "abc",
    "nextPage": "slksfd",
  },
}
`;
